doc/user: Refresh the Getting Started guide (#18859)
MaterializeInc/materialize: doc/user/content/get-started.md

  - /install/
---

> This guide walks you through getting started with Materialize, including:

Reviewer: Something we may want to consider - direct action language:
```
Welcome to the Materialize Getting Started guide. In this guide, you will:

- Connect...
- Compute...
- Simulate...
```

  - /install/
---

This guide walks you through getting started with Materialize.

- First, you'll sign up for a free Materialize trial account and prepare your environment.
- Next, you'll connect a sample streaming data source and explore the unique value of Materialize: computing real-time results over fast-changing data.

> - Finally, you'll simulate a failure to see active replication in action.

Reviewer 2: In the previous version, the intro served more the purpose of quickly highlighting some key concepts, as well as features that users frequently ask about, like joins and time windows (_i.e._ temporal filters, not window functions). This feels slightly less useful.

Author: Yeah, this new version summarizes the flow and then mentions some concepts you'll learn through the experience, but I also don't love it. I'll iterate on this, but the previous version didn't really signal anything about the "wow" of the product, which I'd like to find a way to keep, as that's the primary experience we want users to take-away.

Along the way, you'll learn about basic concepts like clusters and replicas and how SQL features like views, indexes, and materialized views work in the context of Materialize. 

## Before you begin 

Materialize is wire-compatible with PostgreSQL, which means it integrates with most client libraries, ORM frameworks, and other [third-party tools](/integrations/) that support PostgreSQL.

In this guide, you'll use [`psql`](https://www.postgresql.org/docs/current/app-psql.html) to interact with Materialize, so make sure you have it installed locally.

{{< tabs >}}

> {{< tab "Mac">}}

Reviewer 3: Nit: `Mac` is the brand of computer hardware; `macOS` is the OS it runs. Since we're saying `Linux` and `Windows` below (as opposed to Thinkpad, Dell, etc.) I think we should also say `macOS` here for consistency.

{{< /tab >}}
{{< /tabs >}}

> ## Step 1. Start a free trial

Reviewer: Can we add an anchor link to the next step for those users?
> If you already have a Materialize account, skip to the [next section](#-link)

> ## Step 1. Start a free trial

Author: Good call, @tr0njavolta. Added that.

With a free trial, you get 14 days of access to Materialize resources worth up to [4 credits per hour](https://materialize.com/pricing/). This limit should accommodate most trial scenarios. However, if you need additional time or resources, reach out to sales@materialize.com.      

```sql
CREATE CLUSTER quickstart REPLICAS (small_replica (SIZE = '2xsmall'));
```
1. [Sign up for a Materialize trial account](https://materialize.com/register/) using your business email address. 

And make it our active cluster:
1. Check your inbox for a confirmation email from Materialize and follow the instructions to activate your account.

```sql
SET cluster = quickstart;
```
    In most cases, your confirmation should arrive in 30 minutes or less.

Now that **physical isolation** is out of the way, let's also take care of **logical isolation** and create a new schema to make sure we don't run into [namespacing](/sql/namespaces/) annoyances (like naming collisions):
## Step 2. Prepare your environment

```sql

> CREATE SCHEMA qck;

Reviewer 4: Schemas are used for a selfish purpose in the current get started: to differentiate in our analytics which objects were created as part of the demo from others. There's no reason a user has work around this for us, but we should consider some other object naming conventions (like a prefix) we can reliably assume are "ours" and not a user's organic ones

Author: For now, going with a `_qck` suffix.

> In Materialize, a [cluster](/overview/key-concepts/#clusters) is an **isolated workspace**, similar to a database in PostgreSQL or a virtual warehouse in Snowflake. Within a cluster, you have [replicas](/overview/key-concepts/#cluster-replicas), which are the **physical resources** for doing computational work. Clusters are completely isolated from each other, so replicas can be sized based on the specific task of the cluster, whether that is ingesting data from a source, computing always-up-to-date query results, serving results to clients, or a combination.

Reviewer 3: The analogy to databases in Postgres is slightly imprecise, since catalog objects are not scoped to a cluster in Materialize (e.g. views and tables are visible from all clusters). Also, you can query a MV or source maintained by one cluster from another cluster, whereas IIRC (could be wrong) cross-database queries aren't possible in Postgres.

Author: Removed the postgres mention here, and changed `workspace` to `environment`.

> For this guide, you'll create 2 clusters, each with a single replica. One cluster will be for ingesting source data from Kafka. The other will be for computing and serving query results.

Author: Done.

To ingest data from our Kafka cluster, there are three Materialize objects you need to work with: 
[Secrets](/sql/create-secret/) for storing access credentials for the external system; [connections](/sql/create-connection/#kafka) for describing how to connect and authenticate; and [sources](/sql/create-source/) for defining the Kafka topics to use. 

1. [Click here](https://cloud.materialize.com/showSourceCredentials) to get the access credentials for our Kafka cluster. 

    You'll see pre-formatted [`CREATE SECRET`](/sql/create-secret/) commands including the credentials. They'll look something like:
    
    ```sql
    CREATE SECRET kafka_user AS '<KAFKA-USER>';
    CREATE SECRET kafka_password AS '<KAFKA-PASSWORD>';
    CREATE SECRET csr_user AS '<CSR-USER>';
    CREATE SECRET csr_password AS '<CSR-PASSWORD>';
    ```

    Copy the commands and run them in the SQL shell to create your secrets. 
     

> 1. Use the [`CREATE CONNECTION`](/sql/create-connection/#kafka) command to create a connection between Materialize and the Kafka cluster, using the secrets from the previous step for [SASL authentication](/sql/create-connection/#sasl-t0):

Author: Done.

For this guide, you'll create 2 clusters, one for ingesting source data and the other for computing and serving query results. Each cluster will contain a single replica at first (you'll explore the value of adding replicas later). 

1. In the [Materialize UI](https://console.materialize.com/), enable the region where you want to run Materialize. 

    Region setup will take a few minutes. 

1. On the **Connect** screen, create a new app password and then copy the `psql` command. 

    The app password will be displayed only once, so be sure to copy the password somewhere safe. If you forget your password, you can create a new one. 

1. Open a new terminal window, run the `psql` command, and enter your app password.

    In the SQL shell, you'll be connected to a [pre-installed `default` cluster](/sql/show-clusters/#pre-installed-clusters) from which you can get started.

1. Use the [`CREATE CLUSTER`](/sql/create-cluster/) command to create two new clusters, each with a single replica:

    ```sql

>     CREATE CLUSTER ingest_qck REPLICAS (r1 (SIZE = '2xsmall'));

Reviewer 2: Oh, we could totally rename this to `Quickstart`. Just need to be careful and add an alias so existing links to the page don't break.

>     CREATE CLUSTER ingest_qck REPLICAS (r1 (SIZE = '2xsmall'));

Author: Going to leave this for now.

     compute_qck | r3      | xsmall  | t
    (2 rows)
    ```

    Once you see `t` for `ready`, move on to the next step.

## Step 3. Ingest streaming data

Materialize supports streaming data from multiple external sources, including [Kafka](/sql/create-source/kafka/) and [PostgreSQL](/sql/create-source/postgres/). The process for integrating a source typically involves configuring the source's network and creating connection objects in Materialize. 

For this guide, you'll keep it simple and use a [built-in load generator](/sql/create-source/load-generator/#auction) that simulates an auction house, where users bid on an ongoing series of auctions. 

1. Use the [`CREATE SOURCE`](/sql/create-source/load-generator/) command to create the auction house source: 

    ```sql
    CREATE SOURCE auction_house
      IN CLUSTER ingest_qck
      FROM LOAD GENERATOR AUCTION (TICK INTERVAL '50ms')
      FOR ALL TABLES;
    ```

    Note that the `IN CLUSTER` clause attaches this source to the existing `ingest_qck` cluster, but it's also possible to create a cluster and replica at the time of source creation using the [`WITH SIZE`](/sql/create-source/load-generator/#with-options) option.

> 1. With the source created, Materialize starts ingesting data into durable storage, automatically splitting the stream into multiple "subsources" that represent different tables/topics. Use the [`SHOW SOURCES`](/sql/show-sources/) command to take a look:

Author: Done.

    ```sql
    SHOW SOURCES;
    ```
    <p></p>

    ```nofmt
              name          |      type      | size
    ------------------------+----------------+------
     accounts               | subsource      |
     auction_house          | load-generator |
     auction_house_progress | subsource      |
     auctions               | subsource      |
     bids                   | subsource      |
     organizations          | subsource      |
     users                  | subsource      |
    (7 rows)
    ```

    In addition to the `auction_house` load generator source and its subsources, you'll see `auction_house_progress`, which is created by Materialize so you can [monitor source ingestion](/sql/create-source/load-generator/#monitoring-source-progress). 
    
    <!-- NOT SURE IF THIS IS CORRECT: In production scenarios, it's important to note that Materialize takes an [initial snapshot](/ops/troubleshooting/#has-my-source-ingested-its-initial-snapshot) of a source that must complete before Materialize can guarantee correct results.   -->

> 1. Before moving on, look at the schema of the tables you'll be working with:

Author: Done.

    ```sql
    SHOW COLUMNS FROM auctions;
    ```
    <p></p>

    ```nofmt
       name   | nullable |           type
    ----------+----------+--------------------------
     id       | f        | bigint
     seller   | f        | bigint
     item     | f        | text
     end_time | f        | timestamp with time zone
    (4 rows)
    ```    

    ```sql
    SHOW COLUMNS FROM bids;
    ```
    <p></p>

    ```nofmt
        name    | nullable |           type
    ------------+----------+--------------------------
     id         | f        | bigint
     buyer      | f        | bigint
     auction_id | f        | bigint
     amount     | f        | integer
     bid_time   | f        | timestamp with time zone
    (5 rows)
    ```

## Step 4. Compute real-time results

With auction data streaming in, you can now explore the unique value of Materialize: computing real-time results over fast-changing data.  

1. Switch to your `compute_qck` cluster:

    ```sql
    SET CLUSTER = compute_qck;
    ```

1. Enable `psql`'s timing feature so you can see how quickly results are returned:

    ```sql
    \timing
    ```

> 1. Let's start by creating a [**non-materialized view**](/overview/key-concepts/#non-materialized-views):

Author: Good point. Fixed. I wasn't aware that view + index used to mean "materialized view".

    ```sql
    CREATE VIEW avg_bids AS 
      SELECT auctions.item, avg(bids.amount) AS average_bid
      FROM bids
      JOIN auctions ON bids.auction_id = auctions.id
      WHERE bids.bid_time < auctions.end_time
      GROUP BY auctions.item;
    ```
    
    This view joins data from `auctions` and `bids` to get the average price of bids eligible to be winners, that is, bids that arrived before their auctions closed. 
    

>     Note that, as in other SQL databases, a non-materialized view in Materialize is simply an alias for the embedded `SELECT` statement. The results of the query are computed only when the view is queried.

Author: Done.

    ```

     ```sql
    SELECT * FROM avg_bids;
    ```
    <p></p>

    ```nofmt
            item        |    average_bid
    --------------------+--------------------
     Custom Art         |  49.76620397600282
     Gift Basket        | 49.850028105677346
     City Bar Crawl     |  50.08233974737339
     Best Pizza in Town |  50.46824567514223
     Signed Memorabilia |  50.12977674688315
    (5 rows)

    Time: 846.322 ms
    ```

    Latency is high again because the index you created on the view exists only inside the `compute_qck` cluster. In the `default` cluster, where you are currently, you don't have access to the index's pre-computed results. Instead, the view is again retrieving data from durable storage and computing the results at query-time.  

1. There are lots of cases where you'll want results to be accessible from multiple clusters, however. In such cases, you use [materialized views](/overview/key-concepts/#materialized-views). 

    Like an index, a materialized view incrementally computes the results of a query as new data arrives. But unlike an index, **a materialized view persists its results to durable storage** that is accessible to all clusters. 

>     From within the `default` cluster, create a materialized view to see this in action:

Author: Good idea! Done.

## Step 5. Survive failures

Earlier, when you created your clusters, you gave each cluster one [replica](/overview/key-concepts/#cluster-replicas), that is, one physical resource. For the `ingest_qck` cluster, that's the max number of replicas allowed, as clusters for sources can have only one replica. For the `compute_qck` cluster, however, you can increase the number of replicas for greater tolerance to replica failure. 

Each replica in a non-source cluster is a logical clone, doing the same computation and holding the same results in memory. This design provides Materialize with active replication, and so long as one replica is still reachable, the cluster continues making progress. 

Let's see this in action.

1. Add a second replica to the `compute_qck` cluster:

    ```sql
    CREATE CLUSTER REPLICA compute_qck.r2 SIZE = '2xsmall';
    ```

1. Check the status of the new replica:

    ```sql
    SHOW CLUSTER REPLICAS WHERE CLUSTER = 'compute_qck';
    ```
    <p></p>

    ```noftm
       cluster   | replica |  size   | ready
    -------------+---------+---------+-------
     compute_qck | r1      | 2xsmall | t
     compute_qck | r2      | 2xsmall | t
    (2 rows)
    ```

> 1. Once the new `r2` replica is ready, drop the `r1` replica to simulate a failure:

Author: Done.

     Signed Memorabilia |  50.11854192264935
    (5 rows)

    Time: 23.537 ms
    ```

    As you can see, the results are available despite the failure of one of the cluster's replicas.

## Step 6. Scale up or down 

In addition to using replicas to increase fault tolerance, you can add and remove replicas to scale resources up or down according to the needs of a cluster. For example, let's say the `2xsmall` replica in the `compute_qck` cluster is running low on memory. 

1. Add the next largest replica, `xsmall`: 

    ```sql
    CREATE CLUSTER REPLICA compute_qck.r3 SIZE = 'xsmall';
    ```

1. Use the [`SHOW CLUSTER REPLICAS`](https://materialize.com/docs/sql/show-cluster-replicas/) command to check the status of the new replica:

    ```sql
    SHOW CLUSTER REPLICAS WHERE CLUSTER = 'compute_qck';
    ```
    <p></p>

    ```noftm
       cluster   | replica |  size   | ready
    -------------+---------+---------+-------
     compute_qck | r2      | 2xsmall | t
     compute_qck | r3      | xsmall  | t
    (2 rows)
    ```

1. Once the `r3` replica is ready, it's safe to remove the `r2` replica:

    ```sql
    DROP CLUSTER REPLICA compute_qck.r2;
    ```

> 1. Again, because replicas in a cluster are logical clones, the new replica returns results just like old replica:

Author: Done.

    ```sql
    SELECT * FROM avg_bids;
    ```
    <p></p>

    ```nofmt
            item        |    average_bid
    --------------------+-------------------
     Custom Art         | 49.82171985815603
     Gift Basket        | 49.81225672519678
     City Bar Crawl     | 50.16030259365994
     Best Pizza in Town | 50.32564214192425
     Signed Memorabilia | 49.96557507282196
    (5 rows)

    Time: 31.655 ms
    ```

## What's next?

You just created your first transformations on streaming data using common SQL patterns, and tried to break Materialize! We encourage you to continue exploring the sample auction house source, or try one of the [use-case-specific quickstarts](https://github.com/MaterializeInc/materialize/tree/main/doc/user/content/quickstarts). 

If you’re done for the time being, remember to clean up your environment:

```sql
DROP CLUSTER ingest_qck CASCADE;
```

```sql
DROP CLUSTER compute_qck CASCADE;
```

> For a more comprehensive overview of the basic concepts behind Materialize, read through [What is Materialize?](/overview/what-is-materialize)

Author: Since the link ends with a question mark, adding a final period feels grammatically awkward.

> For a more comprehensive overview of the basic concepts behind Materialize, read through [What is Materialize?](/overview/what-is-materialize)

Reviewer 2: Ah, I see!

{{< /tab >}}
{{< /tabs >}}

## Step 1. Start a free trial

With a free trial, you get 14 days of access to Materialize resources worth up to [4 credits per hour](https://materialize.com/pricing/). This limit should accommodate most trial scenarios. However, if you need additional time or resources, reach out to sales@materialize.com.

1. [Sign up for a Materialize trial](https://materialize.com/register/) using your business email address.

1. Activate your Materialize account.

    Once your account is ready, you'll receive an email from Materialize asking you to activate your account. In the process, you'll create credentials for logging into the Materiale UI.

## Step 2. Prepare your environment

> In Materialize, a [cluster](/overview/key-concepts/#clusters) is an **isolated environment**, similar to a virtual warehouse in Snowflake. Within a cluster, you have [replicas](/overview/key-concepts/#cluster-replicas), which are the **physical resources** for doing computational work. Clusters are completely isolated from each other, so replicas can be sized based on the specific task of the cluster, whether that is ingesting data from a source, computing always-up-to-date query results, serving results to clients, or a combination.

Reviewer: ```suggestion
In Materialize, a [cluster](/overview/key-concepts/#clusters) is an **isolated environment**, similar to a virtual warehouse in Snowflake. Within a cluster, you have [replicas](/overview/key-concepts/#cluster-replicas), which are the **physical resources** for doing computational work. Clusters are completely isolated from each other, so replicas can be sized based on the specific task of the cluster, whether it is ingesting data from a source, computing always-up-to-date query results, serving results to clients, or a combination.
```

Author: Done.

1. Open a new terminal window, run the `psql` command, and enter your app password.

    In the SQL shell, you'll be connected to a [pre-installed `default` cluster](/sql/show-clusters/#pre-installed-clusters) from which you can get started.

1. Use the [`CREATE CLUSTER`](/sql/create-cluster/) command to create two new clusters, each with a single replica:

    ```sql
    CREATE CLUSTER ingest_qck REPLICAS (r1 (SIZE = '2xsmall'));
    ```

    ```sql
    CREATE CLUSTER compute_qck REPLICAS (r1 (SIZE = '2xsmall'));
    ```

    The `2xsmall` replica size is sufficient for the data ingestion and computation in this getting started scenario. Also, these replicas use only 1 of the [4 credits per hour](https://materialize.com/pricing/) you get with your free trial. Combined with the `xsmall` replica in the pre-installed `default` cluster, which uses 1 credit per hour, you still have 2 hourly credits available.

1. Use the [`SHOW CLUSTER REPLICAS`](https://materialize.com/docs/sql/show-cluster-replicas/) command to check the status of the replicas:

    ```sql

>     SHOW CLUSTER REPLICAS WHERE CLUSTER = 'compute_qck';

Author: Hmm, I'm used to this pattern and haven't seen it cause much confusion. I'd prefer to stick with this for now and adjust if we see users having trouble.

    ```

    ```noftm
       cluster   | replica |  size   | ready
    -------------+---------+---------+-------
     compute_qck | r1      | 2xsmall | t
    (2 rows)
    ```

    Once the `r1` replica is ready (`ready=t`), move on to the next step.

## Step 3. Ingest streaming data

Materialize supports streaming data from multiple external sources, including [Kafka](/sql/create-source/kafka/) and [PostgreSQL](/sql/create-source/postgres/). The process for integrating a source typically involves configuring the source's network and creating connection objects in Materialize.

> For this guide, you'll keep it simple and use a [built-in load generator](/sql/create-source/load-generator/#auction) that simulates an auction house, where users bid on an ongoing series of auctions.

Reviewer: ```suggestion
For this guide, you'll use a [built-in load generator](/sql/create-source/load-generator/#auction) that simulates an auction house, where users bid on an ongoing series of auctions.
```

Author: I feel like I need this to create a contrast to the typical process mentioned prior.

1. Use the [`CREATE SOURCE`](/sql/create-source/load-generator/) command to create the auction house source:

    ```sql
    CREATE SOURCE auction_house
      IN CLUSTER ingest_qck
      FROM LOAD GENERATOR AUCTION (TICK INTERVAL '50ms')
      FOR ALL TABLES;
    ```

>     Note that the `IN CLUSTER` clause attaches this source to the existing `ingest_qck` cluster, but it's also possible to create a cluster and replica at the time of source creation using the [`WITH SIZE`](/sql/create-source/load-generator/#with-options) option.

Reviewer: Maybe format as a Note/Aside since it's not directly actionable in this guide.

Author: I would prefer note formatting here and a few other places as well, but I couldn't get it to work. When nested in a step, the note shortcode renders wrong:

<img width="787" alt="Screenshot 2023-04-24 at 4 14 57 PM" src="https://user-images.githubusercontent.com/15893490/234106296-9e4a948d-416d-4c00-a43e-04a00cc2b042.png">

Logged an issue: https://github.com/MaterializeInc/materialize/issues/18933

> 1. With the source created, Materialize starts ingesting data into durable storage, automatically splitting the stream into multiple _subsources_ that represent different tables. Use the [`SHOW SOURCES`](/sql/show-sources/) command to get an idea of the data being generated:

Reviewer: ```suggestion
1. Now that you've created a source, Materialize starts ingesting data into durable storage, automatically splitting the stream into multiple _subsources_ that represent different tables. Use the [`SHOW SOURCES`](/sql/show-sources/) command to view the data being generated:
```

Author: Done.

    ```sql

>     SHOW SOURCES;

Reviewer: Same question around the terminal command indicator, since we're using `psql` to interact with MZ, it might be helpful to new users to differentiate from a CLI and psql command. `materialize=>` for the command prompt

Author: This doesn't work today. Again, would love you to love this as a tooling improvement.

    ```
    <p></p>

    ```nofmt
              name          |      type      | size
    ------------------------+----------------+------
     accounts               | subsource      |
     auction_house          | load-generator |
     auction_house_progress | subsource      |
     auctions               | subsource      |
     bids                   | subsource      |
     organizations          | subsource      |
     users                  | subsource      |
    (7 rows)
    ```

>     In addition to the `auction_house` load generator source and its subsources, you'll see `auction_house_progress`, which is created by Materialize so you can [monitor source ingestion](/sql/create-source/load-generator/#monitoring-source-progress).

Reviewer: ```suggestion
    In addition to the `auction_house` load generator source and its subsources, you'll see `auction_house_progress`, which Materialize creates so you can [monitor source ingestion](/sql/create-source/load-generator/#monitoring-source-progress).
```

Author: Done.

    ```sql
    SHOW COLUMNS FROM auctions;
    ```
    <p></p>

    ```nofmt
       name   | nullable |           type
    ----------+----------+--------------------------
     id       | f        | bigint
     seller   | f        | bigint
     item     | f        | text
     end_time | f        | timestamp with time zone
    (4 rows)
    ```

    ```sql
    SHOW COLUMNS FROM bids;
    ```
    <p></p>

    ```nofmt
        name    | nullable |           type
    ------------+----------+--------------------------
     id         | f        | bigint
     buyer      | f        | bigint
     auction_id | f        | bigint
     amount     | f        | integer
     bid_time   | f        | timestamp with time zone
    (5 rows)
    ```

## Step 4. Compute real-time results

With auction data streaming in, you can now explore the unique value of Materialize: computing real-time results over fast-changing data.

1. Switch to your `compute_qck` cluster:

    ```sql
    SET CLUSTER = compute_qck;
    ```

1. Enable `psql`'s timing feature so you can see how quickly results are returned:

    ```sql
    \timing
    ```

> 1. Let's start by creating a [**view**](/overview/key-concepts/#non-materialized-views):

Reviewer: ```suggestion
1. First, create a [**view**](/overview/key-concepts/#non-materialized-views):
```

Author: Done.

    ```sql
    CREATE VIEW avg_bids AS
      SELECT auctions.item, avg(bids.amount) AS average_bid
      FROM bids
      JOIN auctions ON bids.auction_id = auctions.id
      WHERE bids.bid_time < auctions.end_time
      GROUP BY auctions.item;
    ```

>     This view joins data from `auctions` and `bids` to get the average price of bids eligible to be winners, that is, bids that arrived before their auctions closed.

Reviewer: ```suggestion
    This view joins data from `auctions` and `bids` to get the average price of eligible bids that arrived before their auctions closed.
```

Author: Done, but also removed `eligible` because we're not saying what for.

>     Note that, as in other SQL databases, a view in Materialize is simply an alias for the embedded `SELECT` statement. The results of the query are computed only when the view is queried.

Reviewer: ```suggestion
    Note that, as in other SQL databases, a view in Materialize is an alias for the embedded `SELECT` statement. Materialize computes the results of the query only when you query the view.
```

Author: Done. Thanks for helping to fix some passive voice in here. I think the second case is misleading, though, since it suggests that user needs to be involved, when it can be a client, etc.

1. Query the view a few times:

    ```sql
    SELECT * FROM avg_bids;
    ```
    <p></p>

    ```nofmt
            item        |    average_bid
    --------------------+-------------------
     Custom Art         | 50.10550599815441
     Gift Basket        | 50.51195882531032
     City Bar Crawl     | 50.02785145888594
     Best Pizza in Town | 50.20555741546703
     Signed Memorabilia | 49.34376098418278
    (5 rows)

    Time: 738.987 ms
    ```

    ```sql
    SELECT * FROM avg_bids;
    ```
    <p></p>

    ```nofmt
            item        |    average_bid
    --------------------+--------------------
     Custom Art         | 50.135432589422194
     Gift Basket        | 50.485373134328356
     City Bar Crawl     |  50.03637566137566
     Best Pizza in Town |  50.16190159574468
     Signed Memorabilia | 49.354624781849914
    (5 rows)

    Time: 707.403 ms
    ```

>     You'll see the average bid change as new auction data streams into Materialize. However, the view is retrieving data from durable storage and computing results at query-time, so latency is high and would be much higher with a production dataset.

Reviewer: ```suggestion
    You'll see the average bid change as new auction data streams into Materialize. However, the view retrieves data from durable storage and computes results at query-time, so latency is high and would be much higher with a production dataset.
```

Author: Done.

> 1. Let's now create an [**index**](/overview/key-concepts/#indexes) on the view:

Reviewer: ```suggestion
1. Next, create an [**index**](/overview/key-concepts/#indexes) on the view:
```

Author: Done.

    ```nofmt
            item        |    average_bid
    --------------------+--------------------
     Custom Art         | 49.783986655546286
     Gift Basket        |  49.93436483689761
     City Bar Crawl     |  49.93733653007847
     Best Pizza in Town |  50.43617136074242
     Signed Memorabilia |  50.09202958093673
    (5 rows)

    Time: 26.403 ms
    ```

    You'll see the average bids continue to change, but now that the view is indexed and results are pre-computed and stored in memory, latency is down to 26 milliseconds!

1. One thing to note about indexes is that they exist only in the cluster where they are created. To experience this, switch to the `default` cluster and query the view again:

    ```sql
    SET CLUSTER = default;
    ```

     ```sql
    SELECT * FROM avg_bids;
    ```
    <p></p>

    ```nofmt
            item        |    average_bid
    --------------------+--------------------
     Custom Art         |  49.76620397600282
     Gift Basket        | 49.850028105677346
     City Bar Crawl     |  50.08233974737339
     Best Pizza in Town |  50.46824567514223
     Signed Memorabilia |  50.12977674688315
    (5 rows)

    Time: 846.322 ms
    ```

>     Latency is high again because the index you created on the view exists only inside the `compute_qck` cluster. In the `default` cluster, where you are currently, you don't have access to the index's pre-computed results. Instead, the view is again retrieving data from durable storage and computing the results at query-time.

Reviewer: ```suggestion
    Latency is high again because the index you created on the view exists only inside the `compute_qck` cluster. In the `default` cluster, where you are currently, you don't have access to the index's pre-computed results. Instead, the view once again retrieves data from durable storage and computes the results at query-time.
```

Author: Done.

> 1. There are lots of cases where you'll want results to be accessible from multiple clusters, however. In such cases, you use [materialized views](/overview/key-concepts/#materialized-views).

Reviewer: ```suggestion

1. In many cases, you need multiple clusters to have access to results across your environment. To achieve this, you use [materialized views](/overview/key-concepts/#materialized-views).
```

Author: Done, partially. Given the paragraph prior, I feel like I need to leave some sense of transition, even though this isn't idea in this numbered format.

    SELECT * FROM num_bids;
    ```
    <p></p>

    ```nofmt
            item        | number_of_bids
    --------------------+----------------
     Custom Art         |          10634
     Gift Basket        |          11266
     City Bar Crawl     |          10292
     Best Pizza in Town |          10498
     Signed Memorabilia |          10801
    (5 rows)

    Time: 790.384 ms
    ```

    As you can see, although the materialized view was created in the `default` cluster, its results are available from other clusters as well because they are in shared, durable storage.

1.  If retrieving a materialized view's results from storage is too slow, you can create an index on the materialized view as well:

    ```sql
    CREATE INDEX num_bids_idx ON num_bids (item);
    ```

    ```sql
    SELECT * FROM num_bids;
    ```
    <p></p>

    ```nofmt
            item        | number_of_bids
    --------------------+----------------
     Custom Art         |          14373
     Gift Basket        |          15271
     City Bar Crawl     |          14294
     Best Pizza in Town |          14606
     Signed Memorabilia |          14843
    (5 rows)

    Time: 32.064 ms
    ```

>     Now that the materialzed view's results are served from memory, latency is low again.

Reviewer: ```suggestion
    Now that the materialized view serves results from memory, latency is low again.
```

## Step 5. Survive failures

Earlier, when you created your clusters, you gave each cluster one [replica](/overview/key-concepts/#cluster-replicas), that is, one physical resource. For the `ingest_qck` cluster, that's the max number of replicas allowed, as clusters for sources can have only one replica. For the `compute_qck` cluster, however, you can increase the number of replicas for greater tolerance to replica failure.

Each replica in a non-source cluster is a logical clone, doing the same computation and holding the same results in memory. This design provides Materialize with active replication, and so long as one replica is still reachable, the cluster continues making progress.

> Let's see this in action.

Reviewer: ```suggestion
In this step, you'll add a replica and simulate a cluster replica failure.
```

Author: Transition feels odd without this. We can revisit.

    ```sql
    CREATE CLUSTER REPLICA compute_qck.r2 SIZE = '2xsmall';
    ```

1. Check the status of the new replica:

    ```sql
    SHOW CLUSTER REPLICAS WHERE CLUSTER = 'compute_qck';
    ```
    <p></p>

    ```noftm
       cluster   | replica |  size   | ready
    -------------+---------+---------+-------
     compute_qck | r1      | 2xsmall | t
     compute_qck | r2      | 2xsmall | t
    (2 rows)
    ```

1. Once the `r2` replica is ready (`ready=t`), drop the `r1` replica to simulate a failure:

    ```sql
    DROP CLUSTER REPLICA compute_qck.r1;
    ```

1. Query the indexed view that you created in `compute_qck` earlier:

    ```sql
    SELECT * FROM avg_bids;
    ```
    <p></p>

    ```nofmt
            item        |    average_bid
    --------------------+--------------------
     Custom Art         | 49.770776201263864
     Gift Basket        |   49.8909070204407
     City Bar Crawl     | 50.056635368698046
     Best Pizza in Town |  50.50023551577956
     Signed Memorabilia |  50.11854192264935
    (5 rows)

    Time: 23.537 ms
    ```

    As you can see, the results are available despite the failure of one of the cluster's replicas.

## Step 6. Scale up or down

> In addition to using replicas to increase fault tolerance, you can add and remove replicas to scale resources up or down according to the needs of a cluster. For example, let's say the `2xsmall` replica in the `compute_qck` cluster is running low on memory.

Reviewer: ```suggestion
In addition to using replicas to increase fault tolerance, you can add and remove replicas to scale resources up or down according to the needs of a cluster. For example, if the `2xsmall` replica in the `compute_qck` cluster is running low on memory.
```

Author: This change makes the sentence a fragment. Going to leave it as-is for now.

For this guide, you'll create 2 clusters, one for ingesting source data and the other for computing and serving query results. Each cluster will contain a single replica at first (you'll explore the value of adding replicas later).

1. In the [Materialize UI](https://console.materialize.com/), enable the region where you want to run Materialize.

    Region setup will take a few minutes.

1. On the **Connect** screen, create a new app password and then copy the `psql` command.

    The app password will be displayed only once, so be sure to copy the password somewhere safe. If you forget your password, you can create a new one.

1. Open a new terminal window, run the `psql` command, and enter your app password.

    In the SQL shell, you'll be connected to a [pre-installed `default` cluster](/sql/show-clusters/#pre-installed-clusters) from which you can get started.

1. Use the [`CREATE CLUSTER`](/sql/create-cluster/) command to create two new clusters, each with a single replica:

    ```sql

>     $ CREATE CLUSTER ingest_qck REPLICAS (r1 (SIZE = '2xsmall'));

Reviewer 2: ```suggestion
    CREATE CLUSTER ingest_qck REPLICAS (r1 (SIZE = '2xsmall'));
```

    );
    ```

> 1. Use the [`CREATE SOURCE`](/sql/create-source/kafka/) command to create sources for our Kafka topics, using the connections from the previous step: 

Author: Done.

    ```sql
    CREATE SOURCE items
      IN CLUSTER ingest
      FROM KAFKA CONNECTION kafka_connection (TOPIC 'mysql.shop.items')
      FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTION csr_connection
      ENVELOPE DEBEZIUM;
    ```

    ```sql
    CREATE SOURCE purchases
      IN CLUSTER ingest
      FROM KAFKA CONNECTION kafka_connection (TOPIC 'mysql.shop.purchases')
      FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTION csr_connection
      ENVELOPE DEBEZIUM;
    ```

>     Note that the `IN CLUSTER` option places these sources in the existing `ingest` cluster, but it's also possible to create a cluster and replica at the time of source creation using the [`WITH SIZE`](/sql/create-source/kafka/#with-options) option.

Author: I agree it's not good to diverge from a lot of the examples in the docs, but I just find it a lot trickier to help users build a mental model of clusters and replicas when using `WITH SIZE`.

1. With sources connected, Materialize starts ingesting the Kafka topics and writing their data to durable storage. Use the [`SHOW SOURCES`](/sql/show-sources/) command to list the sources in the SQL shell:

    ```sql
    SHOW SOURCES;
    ```
    <p></p>

    ```nofmt
            name        |   type    | size
    --------------------+-----------+------
     items              | kafka     |
     items_progress     | subsource |
     purchases          | kafka     |
     purchases_progress | subsource |
    (4 rows)
    ```

    In addition to `items` and `purchases`, which mirror the Kafka topics named in your `CREATE SOURCE` commands, you'll see `items_progress` and `purchases_progress`. These "subsources" are created by Materialize so you can [monitor source ingestion progress](/sql/create-source/kafka/#monitoring-source-progress). 
    

>     It's important to note that before an initial snapshot of the data is complete, Materialize won't be able to guarantee correct results. The length of the snapshotting process depends on the size of the source, but for this scenario, the process will complete quickly. 

Author: I don't think I understand how this works with Kafka yet, but with Postgres at least, it seems to me we can't claim to return "correct" results based on a partial snapshot of the upstream tables. It's true that we can say we're correct based on what we've seen, but is that really "correct" from a user's perspective? 

I suspect we need to explain snapshotting, for Postgres and Kafka, and show users how to know when it's complete, and maybe guide them to not rely on results until that's the case, but I'll need help from others to figure out if this is really the case. cc @frankmcsherry

>     It's important to note that before an initial snapshot of the data is complete, Materialize won't be able to guarantee correct results. The length of the snapshotting process depends on the size of the source, but for this scenario, the process will complete quickly. 

Author: I'll comment this out for now.

1. Before moving on, look at a few records from `items` and `purchases` to get a sense of the dataset:

    ```sql
    SELECT * FROM items LIMIT 1;
    ```
    <p></p>

    ```nofmt
      id  |       name       | category  | price  | inventory | inventory_updated_at |      created_at      |     updated_at
    ------+------------------+-----------+--------+-----------+----------------------+----------------------+---------------------
     1024 | health spark     | gadgets   | 462.23 |      1365 | 2023-03-06T09:07:22Z | 2023-03-06T09:07:22Z | 2023-03-06 09:07:22
    (1 rows)
    ```    

    ```sql
    SELECT * FROM purchases LIMIT 1;
    ```
    <p></p>

    ```nofmt
        id    | user_id | item_id | status | quantity | purchase_price | deleted |      created_at      |     updated_at
    ----------+---------+---------+--------+----------+----------------+---------+----------------------+---------------------
     57409536 |    3661 |    1955 |      1 |        5 |         300.65 |       0 | 2023-04-18T22:50:48Z | 2023-04-18 22:50:48
    (1 row)

>     ```

Reviewer 5: [SELECT * FROM <source> LIMIT 1](https://github.com/MaterializeInc/materialize/issues/15962) is currently slow in materialize in confusing ways. It gives an unrepresentative experience for the rest of the system. I think we should either comment on that or just remove this select. 

You can run `show columns from purchases;` if you want to show off the schema.

Author: Thanks. Doing this instead.

>     Note how the sources can be queried just like tables in a typical SQL database. 

Author: The point is to just show the shape of the data, so I'll do this with `show columns` instead.

> <!-- 1. With sources connected, Materialize starts ingesting the Kafka topics and writing their data to durable storage. Before an initial snapshot of the data is complete, Materialize won't be able to guarantee correct results, so let's query Materialize's internal `mz_source_statistics` table to [check the initial snapshotting progress](/ops/troubleshooting/#has-my-source-ingested-its-initial-snapshot) of our sources:

Reviewer 3: Why's this commented out? Is it a stray change?

Author: Sorry, yes, this was stray. Though now I have a related paragraph elsewhere, commented out because I need help understanding this better.

     u10 | 0         | t                  | 2000              | 2000           | 2000              | 401564
    (2 rows)
    ```

    Once `snapshot_committed` is `t` for both sources, move on to the next step. -->

## Step 4. Compute real-time results

With source data streaming in, you can now explore the unique value of Materialize: computing real-time results over fast-changing data.

1. Switch to your `compute` cluster:

    ```sql
    SET CLUSTER = compute;
    ```

1. Enable `psql`'s timing feature so you can see how quickly results are returned:

    ```sql
    \timing
    ```

1. Materialize is receiving a continuous stream of `purchase` updates. Let's create a [view](/overview/key-concepts/#non-materialized-views) joining data from `purchases` and `items` to count the purchases in each item category:

    ```sql
    CREATE VIEW purchases_per_category AS
      SELECT category, count(*)
      FROM items
      JOIN purchases ON items.id = purchases.item_id
      GROUP BY items.category;
    ```

    As in other SQL databases, a view doesn't store the results of the query but simply provides an alias for the embedded `SELECT` statement.

> 1. Run the view a few times:

Author: This seems more confusing to me because we're not running the embedded select in the view but rather running the view. I prefer to leave this as-is for now.

    
1. To keep the results of this view **in memory** and return results much faster, create an [index](/overview/key-concepts/#indexes) on the view:

    ```sql
    CREATE INDEX purchases_per_category_idx ON purchases_per_category (category);
    ```

1. Run the view a few more times:

    ```sql
    SELECT * FROM purchases_per_category;
    ```
    <p></p>

    ```nofmt
     category  | purchases
    -----------+-----------
     doodads   |    200889
     gadgets   |    203121
     widgets   |    193276
     clearance |    215299
    (4 rows)

    Time: 26.855 ms
    ```

    ```sql
    SELECT * FROM purchases_per_category;
    ```
    <p></p>

    ```nofmt
     category  | purchases
    -----------+-----------
     doodads   |    200899
     gadgets   |    203132
     widgets   |    193290
     clearance |    215310
    (4 rows)

    Time: 28.214 ms
    ```

    You'll see the count per category continue to increase, but now that the view is indexed, results are returned in 28 milliseconds instead of 14 seconds!   

>     Regardless of how complex the underlying view definition is, querying an indexed view is computationally free because **the results of an index are pre-computed in memory** as new data arrives. However, indexes are accessible only from the cluster where they're created, so to access results from multiple clusters, you'll need to use a [materialized view](/overview/key-concepts/#materialized-views).

Author: Great feedback, Arjun. Reworked this.

1. Still in the `compute` cluster, create a materialized view that calculates rolling aggregations of purchase details for each item:

    ```sql
    CREATE MATERIALIZED VIEW item_purchases AS 
    SELECT 
        item_id, 
        SUM(quantity) AS items_sold, 
        SUM(purchase_price) AS revenue, 
        COUNT(id) AS orders, 
        MAX(created_at::timestamp) AS latest_order 
    FROM purchases 
    GROUP BY item_id;
    ```

> 1. Switch to the `default` cluster and query the materialized view:

Reviewer 5: Ingest won’t work until cluster unification lands

> 1. Switch to the `default` cluster and query the materialized view:

Reviewer 2: Hum, you can't _create_ a materialized view, but can't you...query one?

> 1. Switch to the `default` cluster and query the materialized view:

Author: I'll try this in a bit, but I don't think you can query from a source cluster.

> 1. Switch to the `default` cluster and query the materialized view:

Reviewer 2: ```
materialize=> select * from buyers_materialized;
ERROR:  cannot execute queries on cluster containing sources or sinks
```

That's right! I wasn't aware of this limitation. The usability really is 💀 until we land multi-purpose clusters.

    ```sql
    SET CLUSTER = default;
    ```
    <p></p>

    ```sql
    SELECT * FROM item_purchases LIMIT 10;
    ```

    ```nofmt
     item_id | items_sold |  revenue  | orders |    latest_order
    ---------+------------+-----------+--------+---------------------
        1024 |       1300 |    600899 |    438 | 2023-04-20 05:52:10
         768 |       1180 |  344465.6 |    401 | 2023-04-20 05:58:10
         513 |       1393 |    234024 |    455 | 2023-04-20 05:55:27
         256 |       1176 | 435108.24 |    390 | 2023-04-20 05:56:38
         512 |       1233 | 150758.91 |    405 | 2023-04-20 05:56:48
        1280 |       1302 | 488822.88 |    421 | 2023-04-20 05:57:11
        1536 |       1183 | 181152.79 |    401 | 2023-04-20 05:58:14
        1792 |       1204 | 203042.56 |    403 | 2023-04-20 05:59:34
           1 |       1205 |  260304.1 |    414 | 2023-04-20 05:58:46
         257 |       1172 | 469010.96 |    383 | 2023-04-20 05:53:31
    (10 rows)

    Time: 413.154 ms
    ```

>     As you can see, the materialized view can be accessed from the `default` cluster as well. This is because, unlike an index, the **the results of a materialized view are persisted in durable storage** and incrementally updated as new data arrives. Because the storage layer is shared across all clusters, materialized views allow you to share data between different environments.

Reviewer 2: > This is because, unlike an index, the **the results of a materialized view are persisted in durable storage** and incrementally updated as new data arrives.

The results of an index are also incrementally updated as new data arrives, so this phrasing can be a bit misleading.

>     As you can see, the materialized view can be accessed from the `default` cluster as well. This is because, unlike an index, the **the results of a materialized view are persisted in durable storage** and incrementally updated as new data arrives. Because the storage layer is shared across all clusters, materialized views allow you to share data between different environments.

Author: Revised.

>     Also note that Materialize can output data to an external [sink](/overview/key-concepts/#sinks) only from a materialized view. It is not possible to output to a sink from a view or index.

Reviewer 2: Not sure I would plug sinks here. The statement is also slightly incorrect, since sinks can be based on sources, tables or materialized views.

Author: Removed mention of sinks.

> ## Step 5. Survive failures

Reviewer 5: Maybe too much for this guide, but I really love showing off disaster recovery with subscribes. You can actually see the records never stop flowing.

Author: I agree using subscribe is cool for this case, but it felt like one too many concepts to bake in, when we can show off fault tolerance by just querying an index. I'm hoping we can feature subscribe in a follow-up use case-based tutorial.

Earlier, when you created your clusters, you gave each cluster one [replica](/overview/key-concepts/#cluster-replicas) (that is, one physical resource). For the `ingest` cluster, that's the max number of replicas allowed, as clusters for sources can have only one replica. For the `compute` cluster, however, you can increase the number of replicas for greater tolerance to replica failure. 

> Each replica in a non-source cluster is a clone, doing the same computation and holding the same results in memory. This design provides Materialize with active replication, and so long as one replica is still reachable, the cluster continues making progress.

Reviewer 6: I would say "is a logical clone", since the replicas can be different sizes (I'd also add that they can be different sizes). This is cool and underrated since not every database can do that - many other systems only do replication where they have to be physically identical as well).

Author: Done. I added a "Scale up or down" step to show one reason to add/remove replicas of different sizes.

    SET CLUSTER = compute;
    ```

1. Add a second replica to the cluster

    ```sql
    CREATE CLUSTER REPLICA compute.r2 SIZE = '2xsmall';
    ```

1. To simulate a failure, drop the cluster's original replica:

    ```sql
    DROP CLUSTER REPLICA compute.r1;
    ```

1. Run the indexed view that you created in `compute` earlier:

    ```sql
    SELECT * FROM purchases_per_category;
    ```
    <p></p>

    ```nofmt
     category  | purchases
    -----------+-----------
     doodads   |    200889
     gadgets   |    203121
     widgets   |    193276
     clearance |    215299
    (4 rows)

    Time: 26.855 ms
    ```

    As you can see, the results are still available despite the failure of one of the cluster's replicas.

## What's next?

You just created your first transformations on streaming data using common SQL patterns, and tried to break Materialize! We encourage you to continue exploring the sample Kafka source, or try one of the [use-case specific quickstarts](https://github.com/MaterializeInc/materialize/tree/main/doc/user/content/quickstarts). 

If you’re done for the time being, remember to clean up your environment:

```sql
DROP CLUSTER compute CASCADE;

> DROP ClUSTER ingest CASCADE;

Reviewer 2: ```suggestion
DROP CLUSTER ingest CASCADE;
```

Author: Thanks for catching that. Done.
